src/modules/test/main.c

```diff
@@ -0,0 +1,732 @@
+                {
+                    printf("[FAILURE] Failed to parse JSON payload: %s\n", payloadString);
+                    result = EINVAL;
+                }
+            }
+        }
+
+        if (test->payload != NULL)
+        {
+            if (value != NULL)
+            {
+                if (!json_value_equals(test->payload, value))
+                {
+                    printf("[FAILURE]\n\texpected:\n\t\t%s\n\tactual:\n\t\t%s\n", json_serialize_to_string(test->payload), json_serialize_to_string(value));
+                    result = -1;
+                }
+            }
+            else
+            {
+                printf("[FAILURE]\n\texpected:\n\t\t%s\n\tactual:\n\t\tnull\n", json_serialize_to_string(test->payload));
+                result = -1;
+            }
+        }
+
+        if (test->status != mmiStatus)
+        {
+            printf("[FAILURE] expected status '%d', actual '%d'\n", test->status, mmiStatus);
+            result = -1;
+        }
+    }
+    else if (test->type == DESIRED)
+    {
+        if (test->payload != NULL)
+        {
+            payloadString = json_serialize_to_string(test->payload);
+            payloadSize = (int)strlen(payloadString);
+        }
+
+        mmiStatus = module->set(module->session, test->component, test->object, payloadString, payloadSize);
+
+        if (test->status != mmiStatus)
+        {
+            printf("[FAILURE] expected status '%d', actual '%d'\n", test->status, mmiStatus);
+            result = -1;
+        }
+    }
+    else
+    {
+        printf("[FAILURE] Invalid test step type: %d\n", test->type);
+        result = EINVAL;
+    }
+
+    return result;
+}
+
+long long CurrentMilliseconds()
+{
+    struct timeval time;
+    gettimeofday(&time, NULL);
+    long long milliseconds = time.tv_sec * 1000LL + time.tv_usec / 1000;
+    return milliseconds;
+}
+
+int InvokeRecipe(const char* client, const char* path, const char* bin)
+{
+    int status = 0;
+    int failures = 0;
+    int skipped = 0;
+    int total = 0;
+    long long start = 0;
+    long long end = 0;
+    MANAGEMENT_MODULE* module = NULL;
+    STEP* steps = NULL;
+
+    printf("\n");
+    printf("test recipe: %s\n", path);
+    printf("\n");
+
+    if ((0 == (status = ParseRecipe(path, &steps, &total))))
+    {
+        printf("client: '%s'\n", client);
+        printf("bin: %s\n", bin);
+        printf("\n");
+
+        start = CurrentMilliseconds();
+
+        if (status == 0)
+        {
+            printf("running %d steps...\n", total);
+            printf("\n");
+
+            for (int i = 0; i < total; i++)
+            {
+                STEP* step = &steps[i];
+
+                if (step->delay > 0)
+                {
+                    sleep(step->delay);
+                }
+
+                printf("%d/%d ", i + 1, total);
+
+                switch (step->type)
+                {
+                    case COMMAND:
+                        printf("[COMMAND]\n");
```

Comment:
>printf("[COMMAND]\n"); [](http://example.com/codeflow?start=24&length=22)

Use the logging macros, not printf directly. These printf will log nice to console but will not be in the log file. The CI runs may suffer (when re-adding this in future)

Reply:
There is not log file produced directly by the `moduletest` binary. For CI runs, stdout is redirected to a file:

```
$ ./moduletest blah.json > output.txt
``` 

then that file is uploaded as an artifact.